src/Orchard.Hosting.Web/Extensions/ApplicationBuilderExtensions.cs

```diff
@@ -32,14 +37,49 @@ public static class ApplicationBuilderExtensions
             // Route the request to the correct Orchard pipeline
             builder.UseMiddleware<OrchardRouterMiddleware>();
 
-            // Load controllers
+            
+            // Load controllers and specific dependencies of extensions
             var applicationPartManager = builder.ApplicationServices.GetRequiredService<ApplicationPartManager>();
             var extensionManager = builder.ApplicationServices.GetRequiredService<IExtensionManager>();
+
+            var assemblies = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            var appContext = ProjectContext.CreateContextForEachFramework("").FirstOrDefault();
+            var appExporter = appContext.CreateExporter("Debug");
+
+            foreach (var export in appExporter.GetAllExports())
+            {
+                foreach (var asset in export.CompilationAssemblies)
+                {
+                    assemblies.Add(asset.Name);
+                }
+            }
+
             foreach (var extension in extensionManager.AvailableExtensions())
             {
                 var extensionEntry = extensionManager.LoadExtension(extension);
                 applicationPartManager.ApplicationParts.Add(new AssemblyPart(extensionEntry.Assembly));
-            }
+
+                if (!extensionEntry.Assembly.FullName.StartsWith("Orchard.Core")) {
```

Comment:
I think we should get rid of Orchard.Core too... it bread to many cross referenced components in O1... so we should split it for O2